src/geode/geometry/rotation.cpp

```diff
@@ -0,0 +1,126 @@
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+#include <geode/geometry/rotation.h>
+
+namespace
+{
+    std::array< std::array< double, 4 >, 4 > rotation_matrix(
+        const geode::Vector< 4 >& quaternion )
+    {
+        std::array< std::array< double, 4 >, 4 > matrix;
+        matrix[0][0] =
+            1
+            - 2.0
+                  * ( quaternion.value( 1 ) * quaternion.value( 1 )
+                      + quaternion.value( 2 ) * quaternion.value( 2 ) );
+        matrix[1][0] = 2.0
+                       * ( quaternion.value( 0 ) * quaternion.value( 1 )
+                           + quaternion.value( 2 ) * quaternion.value( 3 ) );
+        matrix[2][0] = 2.0
+                       * ( quaternion.value( 2 ) * quaternion.value( 0 )
+                           - quaternion.value( 1 ) * quaternion.value( 3 ) );
+        matrix[3][0] = 0.0;
+
+        matrix[0][1] = 2.0
+                       * ( quaternion.value( 0 ) * quaternion.value( 1 )
+                           - quaternion.value( 2 ) * quaternion.value( 3 ) );
+        matrix[1][1] =
+            1
+            - 2.0
+                  * ( quaternion.value( 2 ) * quaternion.value( 2 )
+                      + quaternion.value( 0 ) * quaternion.value( 0 ) );
+        matrix[2][1] = 2.0
+                       * ( quaternion.value( 1 ) * quaternion.value( 2 )
+                           + quaternion.value( 0 ) * quaternion.value( 3 ) );
+        matrix[3][1] = 0.0;
+
+        matrix[0][2] = 2.0
+                       * ( quaternion.value( 2 ) * quaternion.value( 0 )
+                           + quaternion.value( 1 ) * quaternion.value( 3 ) );
+        matrix[1][2] = 2.0
+                       * ( quaternion.value( 1 ) * quaternion.value( 2 )
+                           - quaternion.value( 0 ) * quaternion.value( 3 ) );
+        matrix[2][2] =
+            1
+            - 2.0
+                  * ( quaternion.value( 1 ) * quaternion.value( 1 )
+                      + quaternion.value( 0 ) * quaternion.value( 0 ) );
+        matrix[3][2] = 0.0;
+
+        matrix[0][3] = 0.0;
+        matrix[1][3] = 0.0;
+        matrix[2][3] = 0.0;
+        matrix[3][3] = 1.0;
+
+        return matrix;
+    }
+} // namespace
+
+namespace geode
+{
+    Vector3D opengeode_geometry_api rotate(
+        const Vector3D& vector, const Vector3D& axis, double angle )
```

Comment:
Could it be a Point3D we rotate around a Vector3D? Is there a reason to enforce a Vector?